protocols/kad/src/handler.rs

```diff
@@ -712,44 +707,68 @@ impl ConnectionHandler for Handler {
+                    let event = if status.supported {
+                        HandlerEvent::ProtocolConfirmed {
+                            endpoint: self.endpoint.clone(),
+                        }
+                    } else {
+                        HandlerEvent::ProtocolNotSupported {
+                            endpoint: self.endpoint.clone(),
+                        }
+                    };
 
-                return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(event));
+                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(event));
+                }
+                _ => {}
             }
-            _ => {}
-        }
 
-        if let Poll::Ready(Some(event)) = self.outbound_substreams.poll_next_unpin(cx) {
-            return Poll::Ready(event);
-        }
+            match self.outbound_substreams.poll_unpin(cx) {
+                Poll::Ready((query, Ok(Ok(Some(response))))) => {
+                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                        process_kad_response(response, query),
+                    ))
+                }
+                Poll::Ready((_, Ok(Ok(None)))) => {
+                    continue;
+                }
+                Poll::Ready((query_id, Ok(Err(e)))) => {
+                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                        HandlerEvent::QueryError {
+                            error: HandlerQueryErr::Io(e),
+                            query_id,
+                        },
+                    ))
+                }
+                Poll::Ready((query_id, Err(_timeout))) => {
+                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                        HandlerEvent::QueryError {
+                            error: HandlerQueryErr::Io(io::ErrorKind::TimedOut.into()),
+                            query_id,
+                        },
+                    ))
+                }
+                Poll::Pending => {}
+            }
 
-        if let Poll::Ready(Some(event)) = self.inbound_substreams.poll_next_unpin(cx) {
-            return Poll::Ready(event);
-        }
+            if let Poll::Ready(Some(event)) = self.inbound_substreams.poll_next_unpin(cx) {
+                return Poll::Ready(event);
+            }
 
-        let num_in_progress_outbound_substreams =
-            self.outbound_substreams.len() + self.num_requested_outbound_streams;
-        if num_in_progress_outbound_substreams < MAX_NUM_SUBSTREAMS
-            && self.num_requested_outbound_streams < self.pending_messages.len()
-        {
-            self.num_requested_outbound_streams += 1;
-            return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
-                protocol: SubstreamProtocol::new(self.protocol_config.clone(), ()),
-            });
-        }
+            if (self.outbound_substreams.len() + self.pending_streams.len()) < MAX_NUM_STREAMS {
```

Comment:
Why isn't enough only `self.outbound_substreams.len()` here?

Reply:
If we don't take into account, how many streams we've requested so far we might exceed the limit once they all resolve and have to drop some of them.

Reply:
I understand that this check is related to the limit, but my question is about why do you add these two `len`?
Really, `pending_streams` contains `sender`'s, `outbound_substreams` contains, as I can see, `receiver`'s. 
So, both `outbound_substreams` and `pending_streams` are related to outbound streams.
If `MAX_NUM_STREAMS` includes outbound and inbound streams, then you should control the length of `inbound_substreams` and `outbound_streams`, isn't it?

Reply:
I think @dgarus is right here. The current approach, adding `outbound_substreams.len()` and `pending_streams`, counts requests without a matching stream twice. Are we missing something @thomaseizinger?

Reply:
Ah yes, you are right. I was still stuck in the "old world" where the not yet received streams had to be accounted for as well! Another benefit of this new pattern, yey! :)